### PR TITLE
Basic simulator with websockets

### PR DESCRIPTION
**Pull request by rummanwaqar**
Simulator uses websockets for communication (127.0.0.1:8765).
The simulator uses simple interpolation to move boxes around.

The client allows testing by reading or writing. 

## Deps
* REQUIRES PYTHON 3.7+ (I am using 3.7.3). 
* Python library are included in `requirements.txt`

## Run:
* Run simulator + server with `./main.py`
* Run client with `./client <write|read>`

### CODE REVIEW

File: python_simulator/socket_server.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        print(f'Client connected: {websocket.remote_address[0]}:{websocket.remote_address[1]}')
+        if self._on_connect is not None:
+            await websocket.send(self._on_connect(websocket.local_address))
+        self._clients.add(websocket)
+        try:
+            async for message in websocket:
+                if self._on_receive is not None:
+                    self._on_receive(message, websocket.remote_address)
+        except Exception:
+            pass
```
Comment (megalphian) [inline python_simulator/socket_server.py]: Is it okay to not log these exceptions? Not a concern now, but could be useful later on in the project

File: python_simulator/block_simulator.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    def get_state(self):
+        return "state"
+
+    def move_block(self, block_id, goal):
+        # find block
+        block = self._get_block_by_id(block_id)
+        if block is None:
+            raise RuntimeWarning(f'block: {block_id} not found')
+
+        self._goal_queue.put_nowait(MoveTask(block, goal, self._move_speed*self._time_step))
```
[the remark this answers is not in the document]
Comment (rummanwaqar) [reply python_simulator/block_simulator.py]: I agree. Will update this if more tasks come up

File: python_simulator/main.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    comm = await SocketServer.create(on_receive=on_receive, on_connect=on_connect)
+    
+    while True:
+        # get game state and publish as json
+        output_msg = {
+            "type": "game_state",
+            "data": [block.get_object() for block in sim.blocks]
+        }
+        comm.send_data(json.dumps(output_msg))
+        await asyncio.sleep(0.1)
```
[the remark this answers is not in the document]
Comment (rummanwaqar) [reply python_simulator/main.py]: agreed. Will update in next version along with IP address + port

File: python_simulator/socket_server.py (hunk 1 above)
Comment (rummanwaqar) [reply python_simulator/socket_server.py]: As discussed does not apply

File: python_simulator/block_simulator.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        '''
+        while True:
+            # wait for goal and validate goal
+            if self._current_goal is None:
+                self._current_goal = await self._goal_queue.get()
+                print(f'Starting task: {self._current_goal.id}')
+
+                # check if block is movable (nothing should be above it)
+                if not self._is_block_movable(self._current_goal.block):
+                    self._on_complete(self._current_goal.id, self._current_goal.block.id, False, 'Block is not movable')
```
[the remark this answers is not in the document]
Comment (rummanwaqar) [reply python_simulator/block_simulator.py]: fixed :)

File: python_simulator/block_simulator.py [hunk 5]
```diff
@@ -113,47 +115,50 @@ def move_block(self, block_id, goal):
...
 
             if self._current_goal is not None:
                 await asyncio.sleep(self._time_step)
                 if self._current_goal.move():
-                    self._on_complete(self._current_goal.id, self._current_goal.block.id, True, '')
-                    self._current_goal = None
-                    self._goal_queue.task_done()
+                    self._trigger_goal_complete(True)
+
+    def _trigger_goal_complete(self, success, error_msg=''):
```
Comment (megalphian) [inline python_simulator/block_simulator.py]: https://giphy.com/embed/qPVzemjFi150Q

**Pull request merged**